Experimental support for the SARIF 2.1.0 report

KengoTODA opened the pull request:
This PR introduce an experimental `BugReporter` implementation which generates report in SARIF 2.1.0 format.
refs https://github.com/spotbugs/discuss/issues/95


### Note for reviewers

* Apply [patch.txt](https://github.com/spotbugs/spotbugs/files/4848835/patch.txt) to try SNAPSHOT version in your local.
* Check [main.sarif](https://github.com/spotbugs/spotbugs/files/4857510/main.sarif.txt) as sample output.

### Known issues

I will find solution for following issues in another PR.

* `version` is empty, even though SpotBugs should read it from `META-INF/MANIFEST.MF`.
* The spotbugs-gradle-plugin cannot use Worker API (cannot find `SarifException` class from its CLASSPATH), so feature flag is necessary like: `./gradlew spotbugs:spotbugsMain -Pcom.github.spotbugs.snom.worker=false`
* `messageString` contains placeholders like `{2.givenClass}` that cannot be handled by SARIF client.
* Need to find how to add location data to `frames` in `exception`.
* Tell where the `srcRoot` is.

ThrawnCA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/BugCollectionAnalyser.java:
```diff
@@ -0,0 +1,59 @@
+    }
+
+    JSONArray getRules() {
+        JSONArray array = new JSONArray();
```
There is a JSONArray constructor that takes a collection directly; why not use it?

ThrawnCA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Location.java:
```diff
@@ -0,0 +1,214 @@
+ */
+class Location {
+    @NonNull
+    private final List<PhysicalLocation> physicalLocations;
```
Why make this a list when we only ever add one element?

ThrawnCA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Message.java:
```diff
@@ -0,0 +1,31 @@
+
+    JSONObject toJSONObject() {
+        JSONObject jsonObject = new JSONObject().put("id", "default");
+        if (!arguments.isEmpty()) {
```
No need for this condition when we're about to iterate over it anyway. A null check might make sense, except that the field is non-null.

ThrawnCA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Notification.java:
```diff
@@ -0,0 +1,49 @@
+    static Notification fromError(@NonNull AbstractBugReporter.Error error) {
+        String id = String.format("spotbugs-error-%d", error.getSequence());
+        Throwable cause = error.getCause();
+        if (cause != null) {
```
"if not...else" is a double negative, please rearrange.

ThrawnCA commented on spotbugs/src/main/java/edu/umd/cs/findbugs/sarif/Result.java:
```diff
@@ -0,0 +1,32 @@
+
+    JSONObject toJSONObject() {
+        JSONObject result = new JSONObject().put("ruleId", ruleId).put("ruleIndex", ruleIndex).put("message", message.toJSONObject());
+        locations.forEach(location -> result.append("locations", location.toJSONObject()));
```
A slightly streamier approach:

    locations.stream().map(Location::toJSONObject).forEach(location -> result.append("locations", location))

lgolding commented:
Hi @KengoTODA, thanks for attaching the sample SARIF output above. I'll take a look. @michaelcfanning FYI.

KengoTODA commented:
PhysicalLocation is not set well, I'll fix it.

KengoTODA commented:
@lgolding I fixed `PhysicalLocation`, and added `tags` and `fullDescription` for rules. Here is new generated SARIF file for reference: [main.sarif.txt](https://github.com/spotbugs/spotbugs/files/4864008/main.sarif.txt)

@ThrawnCA I'm sorry but could you review this again? After I pass you review, I'm going to merge this.
I haven't updated `TextUICommandLine.java` so this PR won't affect spotbugs' behaviour, just add `org.json` to its dependency. I also changed scope of some existing members, but it keeps backward compatibilities.
